### DISCUSSION TITLE
Working on PC / Windows

### DISCUSSION THREAD

**Discussion by dennis-leeyinghui** [ReagentX/imessage-exporter#258] [General]
Hi everyone, I want to say thank you to the creator of imessage-exporter tool @ReagentX.

Also, if you are a novice like me, using a PC / Windows, and would like to use this tool to back up imessages, it's easy:

1. Use [Apple Devices](https://apps.microsoft.com/detail/9np83lwlpz9k?hl=en-US&gl=US) to backup the iPhone
2. Follow the instructions on installation and [execution](https://github.com/ReagentX/imessage-exporter/blob/develop/imessage-exporter/README.md#how-to-use) of `imessage-exporter`
3. Run `imessage-exporter`, using the `-p` flag to point to the root of the backup directory
4. The backed up messages will be exported to the default location (C:\Users\<username>\imessage_export)

**Comment by ReagentX** [comment]
Cheers! You don't actually need to navigate to any folders inside of the iOS backup. From the docs:

```
-p, --db-path <path/to/source>
        Specify an optional custom path for the iMessage database location
        For macOS, specify a path to a `chat.db` file
        For iOS, specify a path to the root of a device backup directory
        If omitted, the default directory is ~/Library/Messages/chat.db
```

Further, you can manually specify the database type, though this is often not necessary:

```
-a, --platform <macOS, iOS>
        Specify the platform the database was created on
        If omitted, the platform type is determined automatically
```

You should only point `imessage-exporter` to the root of an iOS backup, not to the database file itself. Doing so may lead to data loss as it may not be detected as an iOS backup. I have edited your post to reflect this so that other people do not get confused.

You can find more examples of how to run `imessage-exporter` [here](https://github.com/ReagentX/imessage-exporter/blob/develop/imessage-exporter/README.md#examples).

It's worth mentioning that this software should compile to any [Rust target](https://doc.rust-lang.org/rustc/platform-support.html). As long as you can provide the relevant files, you can run an export, regardless of platform.

**Reply by biasply-casual** [reply]
First I want to thank ReagentX for developing this! It's a life saver for my situation.
I'm also a novice coder and have managed to get to a certain point in getting this to work on Windows (10), but I've hit the following error when running ...  

```
C:\Users\STEVE>imessage-exporter -f html -p C:\Users\STEVE\AppData\Roaming\Apple Computer\MobileSync\Backup\dc988cbad65c7b6b65148debc3a7d30895a58abf -o C:\Users\STEVE\Desktop

thread 'main' panicked at C:\Users\STEVE\.cargo\registry\src\index.crates.io-6f17d22bba15001f\imessage-database-1.8.0\src\util\dirs.rs:26:21:
Unable to resolve user home directory: environment variable not found
```

Am I correct in assuming the cargo environment variable path is not set correctly and that I should set it manually, or should I try reinstalling rust/cargo/etc, or, have I given an incorrect command? 

Edit: There appears to be a missing backslash between STEVE and .cargo when I view the post, but when I edit it the backslash is there... I can also confirm it is present in the terminal.  Weird display bug!

**Reply by ReagentX** [reply]
I fixed this in #262 just now! If anyone was interested in why, we called `home()` when the `Options` constructor ran on this line:

https://github.com/ReagentX/imessage-exporter/blob/915cd1a8403eacef2720967b785193c7247b6b62/imessage-exporter/src/app/options.rs#L338

**Reply by biasply-casual** [reply]
LMAO! My timing is perfect!  
I'll un-/re-install and try again.
Thanks!

**Comment by ShiningN1ght** [comment]
After reading, following, reading again, and executing, I've gotten your wonderful application working in Windows 11, except with this error:

> **No HEIC converter found, attachments will not be converted!**

Is there a command-line converter that needs to be in the folder for this to work? I appreciate your work on this.  With a bit of work and common sense, it does for free what so many seemingly identical apps want US$40 or more to do.  As a side-note, I found your utility via reddit/iphone.

**Reply by ShiningN1ght** [reply]
If useful, my command line is: >imessage-exporter.exe -p . -o .\exported-messages --format html

I have the compiled application in the root folder of the backup, for simplicity.  I've tried a couple different versions of the -c, but no difference.

**Reply by ReagentX** [reply]
For `-c compatible` you have to have a supported converter installed on your system. Currently `imessage-exporter` supports `sips` (macOS default) and `convert` ([ImageMagick](https://imagemagick.org)):

https://github.com/ReagentX/imessage-exporter/blob/915cd1a8403eacef2720967b785193c7247b6b62/imessage-exporter/src/app/converter.rs#L33-L45

The error means that we cannot convert files, however `-c efficient` will still copy attachments to your export directory. The error gets logged when you run because `imessage-exporter` checks if it can convert files when you launch it.

***

### Edit: `compatible` and `efficient` have been renamed, see the [documentation](https://github.com/ReagentX/imessage-exporter/tree/develop/imessage-exporter#cross-platform-attachment-conversion) for examples and instructions.

**Reply by ShiningN1ght** [reply]
Thank you, ReagentX.  I haven't used Imagemagick in years.  This should be fun.  Is it still just a drop-in executable?  

EDIT: I have the IM executable.  Does imessage-exporter look for a particular file name for it? Does it need to be in a particular location? Is there an addition launch argument?

**Reply by ReagentX** [reply]
When you install ImageMagick make sure to include the `convert` command: ![image](https://github.com/ReagentX/imessage-exporter/assets/1920666/b26435bc-5b31-465e-be31-f0452ac1761c)

Per the [docs](https://imagemagick.org/script/porting.php#cli):

> The "magick" command is the new primary command of the Shell API, replacing the old "convert" command. 

`imessage-exporter` does not support this at the moment, but it should. I am tracking this in #263.

***

Edit: the above is no longer accurate, #263 was merged in, so the legacy support for `convert` is no longer supported.

**Comment by dexteon** [comment]
Thank you! Is there any way you can add this to the github home for us windows users? It took me awhile to realized i had to WSL to run the cargo commands. Not going to tell you how long it took me to figure out how to get an imessage database lol. Awesome program btw

**Reply by ReagentX** [reply]
You don't need to use WSL, I've tested and used `imessage-exporter` on a vanilla Windows installation. 

What do you mean by GitHub home?

**Comment by TrevorMay** [comment]
Has anyone here had luck getting the tool to export the imessage attachments (like photos and videos) as well?

**Reply by ReagentX** [reply]
Attachments are supported; how are you running the app and what is the source you are extracting the data from?

**Reply by TrevorMay** [reply]
Haha of course right after I comment, I realize I did the same thing above. I pointed it to the db not the root of the backup
Holy cow though thank you for the fast response and for this amazing application!
You're saving me money on iCloud and/or an iMazing license at a time every dollar counts. :)

**Comment by jhow90-blip** [comment]
Hi - Struggling to get this working for myself. My first attempt in life to download a github software and run something. Feeling very overwhelmed tbh...I tried a bunch of stuff a week or two ago and got frustrated and stopped. I am running Windows 11.

Returned to this today and so far, I have done heaps of stuff I don't really comprehend, but today I went to Windows Powershell and typed:

`cargo install imessage-exporter`

I believe an update was published since the last install, so I did it again. A lot of downloaded and compliling things happened. This was the last line at the end of that run:

`Replaced` package `imessage-exporter v3.1.0` with `imessage-exporter v3.2.0` (executable ``imessage-exporter.exe`)`

I then used PowerShell and typed

`imessage-exporter -f txt -p C:\Users\jhow9\Apple\MobileSync -a iOS`

I got this result.

`No HEIC converter found, image attachments will not be converted!
No CAF converter found, audio attachments will not be converted!
No MOV converter found, video attachments will not be converted!
Invalid configuration: Database file `C:\Users\jhow9\Apple\MobileSync\3d/3d0d7e5fb2ce288813306e4d4636395e047a3d28` does not exist at the specified path!`

I previously downloaded the "Apple Devices" program linked above and spent the time doing a backup using that Apple software. So - if I understand correctly, there is now a copy of all my phone's information stored somewhere on my computer. I clicked around and believe that there are gigs of data stored in C:\Users\jhow9\Apple\MobileSync.
Looking through this discussion, I see things about errors with "No HEIC converter found" and references to ImageMagick. So, I a moment ago went to that website and downloaded a .exe and ran the installer and was sure to click on the "Install legacy utilities" mentioned above.

Again I returned to PowerShell and typed:

`imessage-exporter -f txt -p C:\Users\jhow9\Apple\MobileSync -a iOS`
No HEIC converter found, image attachments will not be converted!
No CAF converter found, audio attachments will not be converted!
No MOV converter found, video attachments will not be converted!
Invalid configuration: Database file `C:\Users\jhow9\Apple\MobileSync\3d/3d0d7e5fb2ce288813306e4d4636395e047a3d28` does not exist at the specified path!

Can you give me any guidance on how to proceed from here? 

- How do I know where the data for my messages is stored since that seems to be a problem?
- How do I get imessage-exporter to locate the converter I downloaded?
- Assuming I can overcome these hurdles, and can run the program, how do I know where the extracted text messages will end up so that I can actually go look at them after this works?

**Reply by ReagentX** [reply]
1) `C:\Users\jhow9\Apple\MobileSync` is the path to where all of your backups are stored. You have to provide the path to a specific backup within that directory. 

2) Detail about the converter requirements can be found in the [docs](https://github.com/ReagentX/imessage-exporter/tree/develop/imessage-exporter#cross-platform-attachment-conversion). 

3) If you don't provide an output folder in your command, the default is `C:\Users\jhow9\imessage_export`.

**Comment by boston2029** [comment]
Thanks! Didn't see the readme
